### PR TITLE
fix state summaries

### LINKED ISSUE #5417: datastore wrong when workflow paused?
### Description

State totals wrong, as reported by scanning, in paused workflows.

### Reproducible Example

(I've seen it with a bunch of workflows, however)

```ini
[scheduling]
   [[graph]]
      R1 = "a => b1 & b2 & b3 & b4"
[runtime]
   [[a, b1, b2, b3, b4, b5]]
       script = sleep $((RANDOM % 5))
```

Pause this while `b<m>` are running.  Scan reports several, but not all, of `b<m>` succeeding after the pause.

```console
$ cylc scan -t rich --colour-blind  # after all running tasks have succeeded
bgg/run8 (paused)
    state totals   running:2 succeeded:1
    title          *null*
    description    *null*
    version        8.2.0.dev
    host           NIWA-1022450.niwa.local
    port           43072
```

The GUI gscan sidebar is also wrong.  

`cylc dump -t` and `cylc tui`, and GUI workflow views correctly show no running tasks in the workflow.

Presumably the datastore is not correctly updating state totals data.

### PR DESCRIPTION
**Pull request by dwsutherland**
closes #5417 

**Before**
https://github.com/cylc/cylc-flow/issues/5417#issuecomment-1475583476

**After**
image

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Covered by existing tests.
- [ ] `CHANGES.md` entry included if this is a change that can affect users
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

File: cylc/flow/data_store_mgr.py [hunk 1]
```diff
@@ -1523,18 +1524,13 @@ def prune_data_store(self):
...
-                    if j_id in j_updated:
-                        del j_updated[j_id]
-                del tp_updated[tp_id]
-
             self.deltas[TASK_PROXIES].pruned.append(tp_id)
             self.deltas[JOBS].pruned.extend(node.jobs)
             self.deltas[EDGES].pruned.extend(node.edges)
             parent_ids.add(node.first_parent)
 
+        self.pruned_task_proxies = node_ids
```
[the remark this answers is not in the document]
Comment (dwsutherland) [reply cylc/flow/data_store_mgr.py:1532]: Might be possible ... But can't hurt to change it either way.

[the remark this answers is not in the document]
Comment (dwsutherland) [reply cylc/flow/data_store_mgr.py]: Further down `self.updates_pending` is flagged true if `node_ids` is non-empty, so I put it in there..

[the remark this answers is not in the document]
Comment (oliver-sanders) [reply cylc/flow/data_store_mgr.py]: Cheers.

**Comment by oliver-sanders** [comment]
Reproduced the original issue on 8.1.x, after the pause I saw both running and succeeded tasks in `cylc scan`.

Re-ran the test on this branch, there are no more running tasks, but I still see the succeeded ones. This is much better!

We could do with clearing those succeeded tasks too though, see this screenshot of Tui showing 5 succeeded tasks, but with an empty tree (empty task pool).

Screenshot from 2023-03-22 15-11-18

**Comment by hjoliver** [comment]
Yep, same result here.

**Comment by oliver-sanders** [comment]
(marking as draft to help with prioritisation)

**Comment by dwsutherland** [comment]
Fixed .. The problem was; when the state totals fields are empty they aren't recognized as *set* .. So fields are cleared, and deltas aren't sent.

**Review by wxtim — COMMENTED** [review]
- [x] Code changes make sense (I don't have a deep understanding of this code)
- [x] Functional testing suggests that this works.

Might this be worth a functional test? Happy to write one for you (@dwsutherland) if you are pressed for time and @oliver-sanders also thinks it's worthwhile.

**Comment by oliver-sanders** [comment]
This is hard to functional test as it is timing / internal-state dependent.

An integration test could work though:

```python
with start(schd):  # note scheduler starts in paused mode
    schd.pool.get_task('1', 'a').reset_state(TASK_STATUS_SUCCEEDED)
    assert schd.data_store_mgr.get_items('~a/w//1/a').status == TASK_STATUS_SUCCEEDED

**Comment by dwsutherland** [comment]
> Might this be worth a functional test? Happy to write one for you (@dwsutherland) if you are pressed for time and @oliver-sanders also thinks it's worthwhile. If not, I'll press "approve".

This is covered by some tests:
https://github.com/cylc/cylc-flow/blob/8.1.x/tests/functional/graphql/01-workflow.t
(and many others that ask for summary info no doubt)
Also, the actual code is run with every test that runs a workflow.

The reason for the patch coverage is my handling of potential exceptions (with `try:` `except:` or `if` blocks), that may or may not occur.

**Review by wxtim — APPROVED** [review]

**Review by hjoliver — APPROVED** [review]

**Pull request merged**